File: FinalProject/app/src/main/java/com/example/finalproject/fragments/ChatMiniFragment.java
package com.example.finalproject.fragments;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;
import androidx.fragment.app.FragmentTransaction;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.util.Log;
import android.view.KeyEvent;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;

import com.example.finalproject.R;
import com.example.finalproject.service.Message;
import com.example.finalproject.service.User;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.ValueEventListener;
import com.google.gson.Gson;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.HashMap;

public class ChatMiniFragment extends Fragment {
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_chat_mini, container, false);
    }

    @Override
    public void onViewCreated(@NonNull @NotNull View view, @Nullable @org.jetbrains.annotations.Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        ArrayList<Message> messages = new ArrayList<>();
        DatabaseReference ref = FirebaseDatabase.getInstance().getReference("Message");
        RecyclerView chat = getView().findViewById(R.id.chat_list);
        EditText enterMessage = getView().findViewById(R.id.message);
        Button back = getView().findViewById(R.id.chat_mini_back_button), register = getView().findViewById(R.id.log_out);
        ref.addValueEventListener(new ValueEventListener() {
            @Override
            public void onDataChange(@NonNull @NotNull DataSnapshot snapshot) {
                messages.clear();
                snapshot.getChildren().forEach(e -> messages.add(e.getValue(Message.class)));
                chat.setAdapter(new ChatMiniFragment.ChatAdapter(messages));
                chat.scrollToPosition(messages.size() - 1);
            }

            @Override
            public void onCancelled(@NonNull @NotNull DatabaseError error) {

            }
        });
        back.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                FragmentManager fm = getParentFragmentManager();
                FragmentTransaction fr=fm.beginTransaction();
                fr.remove(fm.findFragmentById(R.id.status));
                fr.commit();
                fr=fm.beginTransaction();
                fr.add(R.id.status, new StatusBarFragment());
                fr.commit();
            }
        });
        chat.setLayoutManager(new LinearLayoutManager(getContext()));
        enterMessage.setOnEditorActionListener(new TextView.OnEditorActionListener() {
            @Override
            public boolean onEditorAction(TextView v, int actionId, KeyEvent event) {
                Message m = new Message();
                m.message = v.getText().toString();
                m.user = new Gson().toJson(MainActivity.player.getUser());
                m.date=new Date().getTime()-Calendar.getInstance().getTimeZone().getOffset(new Date().getTime())*60*1000;
                ref.child(messages.size() + "").setValue(m);
                v.setText("");
                return false;
            }
        });
        chat.scrollToPosition(messages.size() - 1);
    }

    class ChatAdapter extends RecyclerView.Adapter<ChatMiniFragment.ChatAdapter.ChatViewHolder> {

        ArrayList<Message> data;

        public ChatAdapter(ArrayList<Message> messages) {
            data = messages;
        }

        class ChatViewHolder extends RecyclerView.ViewHolder {
            TextView user, message, time;

            public ChatViewHolder(@NonNull @NotNull View itemView) {
                super(itemView);
                user = itemView.findViewById(R.id.user_list);
                time = itemView.findViewById(R.id.time_list);
                message = itemView.findViewById(R.id.message_list);
            }
        }

        @NonNull
        @NotNull
        @Override
        public ChatMiniFragment.ChatAdapter.ChatViewHolder onCreateViewHolder(@NonNull @NotNull ViewGroup parent, int viewType) {
            return new ChatMiniFragment.ChatAdapter.ChatViewHolder(LayoutInflater.from(parent.getContext()).inflate(R.layout.message, parent, false));
        }

        @Override
        public void onBindViewHolder(@NonNull @NotNull ChatMiniFragment.ChatAdapter.ChatViewHolder holder, int position) {
            holder.message.setText(data.get(position).message);
            long time = data.get(position).date / 1000 / 60 +new Date().getTimezoneOffset()*60;
            String mins, hrs;
            mins = time % 60 >= 10 ? time % 60 + "" : "0" + time % 60;
            time /= 60;
            hrs = (time % 24 + 3) % 24 + "";
            String date = hrs + ":" + mins;
            holder.time.setText(date);
            holder.user.setText(new Gson().fromJson(data.get(position).user, User.class).getLogin());
        }

        @Override
        public int getItemCount() {
            return data.size();
        }
    }
}

File: FinalProject/app/src/main/java/com/example/finalproject/fragments/ChatMiniFragment.kt
package com.example.finalproject.fragments

import android.os.Bundle
import android.view.KeyEvent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.finalproject.R
import com.example.finalproject.service.Message
import com.example.finalproject.service.User
import com.google.firebase.database.*
import com.google.gson.Gson
import java.util.*

class ChatMiniFragment : Fragment(), View.OnClickListener, ValueEventListener, TextView.OnEditorActionListener {

    private lateinit var ref:DatabaseReference
    private lateinit var back:Button
    private lateinit var chat:RecyclerView
    private lateinit var enterMessage: EditText
    private val messages = ArrayList<Message>()


    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_chat_mini, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ref = FirebaseDatabase.getInstance().getReference("Message")
        chat= requireView().findViewById(R.id.chat_list)
        enterMessage = requireView().findViewById(R.id.message)
        back = requireView().findViewById(R.id.chat_mini_back_button)
        ref.addValueEventListener(this)
        back.setOnClickListener(this)
        chat.layoutManager = LinearLayoutManager(context)
        enterMessage.setOnEditorActionListener(this)
        chat.scrollToPosition(messages.size - 1)
    }

    private inner class ChatAdapter(var data: ArrayList<Message>) :
        RecyclerView.Adapter<ChatAdapter.ChatViewHolder>() {

        inner class ChatViewHolder(itemView: View) :
            RecyclerView.ViewHolder(itemView) {
            var user: TextView = itemView.findViewById(R.id.user_list)
            var message: TextView = itemView.findViewById(R.id.message_list)
            var time: TextView = itemView.findViewById(R.id.time_list)

        }

        override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ChatViewHolder {
            return ChatViewHolder(
                LayoutInflater.from(parent.context).inflate(R.layout.message, parent, false)
            )
        }

        override fun onBindViewHolder(holder: ChatViewHolder, position: Int) {
            holder.message.text = data[position].message
            var time = data[position].date / 1000 + Calendar.getInstance().timeZone.getOffset(
                Date().time
            )
            val seconds: String = if (time % 60 >= 10) (time % 60).toString() else "0${time % 60}"
            time /= 60
            val minutes: String = if (time % 60 >= 10) (time % 60).toString() else "0${time % 60}"
            time /= 60
            val hours: String = (time % 24).toString()
            val date = "$hours:$minutes:$seconds"
            holder.time.text = date
            holder.user.text = Gson().fromJson(
                data[position].user,
                User::class.java
            ).login
        }

        override fun getItemCount(): Int = data.size
    }

    override fun onClick(p0: View?) {
        if (p0==back){
            val fragmentManager = parentFragmentManager
            var fragmentTransaction = fragmentManager.beginTransaction()
            fragmentTransaction.remove(fragmentManager.findFragmentById(R.id.status)!!)
            fragmentTransaction.commit()
            fragmentTransaction = fragmentManager.beginTransaction()
            fragmentTransaction.add(R.id.status, StatusBarFragment())
            fragmentTransaction.commit()
        }
    }

    override fun onDataChange(snapshot: DataSnapshot) {
        messages.clear()
        for (i in snapshot.children)
            messages.add(i.getValue(Message::class.java)!!)
        chat.adapter = ChatAdapter(messages)
        chat.scrollToPosition(messages.size - 1)
    }

    override fun onCancelled(error: DatabaseError) {
        Toast.makeText(context, error.message, Toast.LENGTH_SHORT).show()
    }

    override fun onEditorAction(p0: TextView?, p1: Int, p2: KeyEvent?): Boolean {
        if (p0==enterMessage){
            val m = Message(
                p0.text.toString(),
                Gson().toJson(MainActivity.player.user),
                Date().time - Calendar.getInstance().timeZone.getOffset(Date().time) * 60L * 1000
            )
            ref.child(messages.size.toString()).setValue(m)
            p0.text = ""
            return true
        }
        return false
    }
}